pkg/landscaper/registry/components/oci.go

```diff
@@ -22,29 +25,59 @@ import (
 
 // ociClient is a component descriptor repository implementation
 // that resolves component references stored in an oci repository.
+// It can also cache local component descriptors.
 type ociClient struct {
 	ociClient ociclient.Client
 	resolver  *cdoci.Resolver
+	cache     []*cdv2.ComponentDescriptor
 }
 
 // NewOCIRegistry creates a new oci registry from a oci config.
-func NewOCIRegistry(log logr.Logger, config *config.OCIConfiguration) (TypedRegistry, error) {
+func NewOCIRegistry(log logr.Logger, config *config.OCIConfiguration, predefinedComponentDescriptors ...*cdv2.ComponentDescriptor) (TypedRegistry, error) {
 	client, err := ociclient.NewClient(log, confighelper.WithConfiguration(config))
 	if err != nil {
 		return nil, err
 	}
 
-	return &ociClient{
-		ociClient: client,
-		resolver:  cdoci.NewResolver().WithOCIClient(client),
-	}, nil
+	return NewOCIRegistryWithOCIClient(client, predefinedComponentDescriptors...)
 }
 
 // NewOCIRegistryWithOCIClient creates a new oci registry with a oci ociClient
-func NewOCIRegistryWithOCIClient(client ociclient.Client) (TypedRegistry, error) {
+// If supplied, it parses and stores predefined component descriptors.
+func NewOCIRegistryWithOCIClient(client ociclient.Client, predefinedComponentDescriptors ...*cdv2.ComponentDescriptor) (TypedRegistry, error) {
+	cache := make([]*cdv2.ComponentDescriptor, 0)
+	for _, cd := range predefinedComponentDescriptors {
+		if cd == nil {
+			continue
+		}
+		cache = append(cache, cd)
+
+		var extractCdFromLabel func(*[]*cdv2.ComponentDescriptor, *cdv2.ComponentDescriptor) error
+		extractCdFromLabel = func(cdList *[]*cdv2.ComponentDescriptor, cd1 *cdv2.ComponentDescriptor) error {
```

Comment:
```suggestion
		extractCdFromLabel := func(cdList *[]*cdv2.ComponentDescriptor, cd1 *cdv2.ComponentDescriptor) error {
```

Reply:
the explicit declaration seems to be required by the compiler as the function is invoked recursively.